src/appleseed/renderer/kernel/rendering/renderercontrollercollection.cpp

```diff
@@ -0,0 +1,118 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header
+#include "renderer/kernel/rendering/renderercontrollercollection.h"
+
+// Standard headers.
+#include <vector>
+
+namespace renderer
+{
+
+//
+// RendererControllerCollection class implementation.
+//
+
+struct RendererControllerCollection::Impl
+{
+    std::vector<IRendererController*> m_renderer_controller;
+};
+
+RendererControllerCollection::RendererControllerCollection()
+  : impl(new Impl())
+{
+}
+
+RendererControllerCollection::~RendererControllerCollection()
+{
+    delete impl;
+}
+
+void RendererControllerCollection::on_rendering_begin()
+{
+    for (auto i : impl->m_renderer_controller)
+        i->on_rendering_begin();
+}
+
+void RendererControllerCollection::on_rendering_success()
+{
+    for (auto i : impl->m_renderer_controller)
+        i->on_rendering_success();
+}
+
+void RendererControllerCollection::on_rendering_abort()
+{
+    for (auto i : impl->m_renderer_controller)
+        i->on_rendering_abort();
+}
+
+void RendererControllerCollection::on_rendering_pause()
+{
+    for (auto i : impl->m_renderer_controller)
+        i->on_rendering_pause();
+}
+
+void RendererControllerCollection::on_rendering_resume()
+{
+    for (auto i : impl->m_renderer_controller)
+        i->on_rendering_resume();
+}
+
+void RendererControllerCollection::on_frame_begin()
+{
+    for (auto i : impl->m_renderer_controller)
+        i->on_frame_begin();
+}
+
+void RendererControllerCollection::on_frame_end()
+{
+    for (auto i : impl->m_renderer_controller)
+        i->on_frame_end();
+}
+
+void RendererControllerCollection::on_progress()
+{
+    for (auto i : impl->m_renderer_controller)
+        i->on_progress();
+}
+
+IRendererController::Status RendererControllerCollection::get_status() const
+{
+    IRendererController::Status status = IRendererController::Status::ContinueRendering;
+    for (auto i : impl->m_renderer_controller)
+        status = i->get_status() > status ? i->get_status() : status;
```

Comment:
That's an interesting choice. You assume that there's an ordering in the statuses but there isn't really one. I think we could live with simply returning the first status that isn't `ContinueRendering`. What do you think?

Reply:
Agreed!